_posts/2021-04-22-torch-for-optimization/torch-for-optimization.Rmd

```diff
@@ -0,0 +1,640 @@
+
+Instead, we let a `torch` optimizer update the candidate $\mathbf{x}$ for us. Habitually, our first try is *Adam*.
+
+### Adam (?)
+
+With Adam, optimization proceeds a lot faster. Truth be told, though, choosing a good learning rate *still* takes non-negligeable experimentation. (Try the default learning rate, 0.001, for comparison.)
+
+```{r}
+num_iterations <- 100
+
+x_star <- torch_tensor(c(-1, 1), requires_grad = TRUE)
+
+lr <- 1
+optimizer <- optim_adam(x_star, lr)
+
+for (i in 1:num_iterations) {
+  
+  if (i %% 10 == 0) cat("Iteration: ", i, "\n")
+  
+  optimizer$zero_grad()
+  value <- rosenbrock(x_star)
+  if (i %% 10 == 0) cat("Value is: ", as.numeric(value), "\n")
+  
+  value$backward()
+  optimizer$step()
+  
+  if (i %% 10 == 0) cat("Gradient is: ", as.matrix(x_star$grad), "\n\n")
+  
+}
+
+```
+
+    Iteration:  10 
+    Value is:  0.8559565 
+    Gradient is:  -1.732036 -0.5898831 
+
+    Iteration:  20 
+    Value is:  0.1282992 
+    Gradient is:  -3.22681 1.577383 
+
+    ...
+    ...
+
+    Iteration:  90 
+    Value is:  4.003079e-05 
+    Gradient is:  -0.05383469 0.02346456 
+
+    Iteration:  100 
+    Value is:  6.937736e-05 
+    Gradient is:  -0.003240437 -0.006630421 
+
+It took us about a hundred iterations to arrive at a decent value. This is a lot faster than the manual approach above, but still quite a lot. Luckily, further improvements are possible.
+
+### L-BFGS (!)
+
+Among the many `torch` optimizers commonly used in deep learning (Adam, AdamW, RMSprop ...), there is one "outsider", much better known in classic numerical optimization than in neural-networks space: L-BFGS, a.k.a. [Limited-memory BFGS](https://en.wikipedia.org/wiki/Limited-memory_BFGS), a memory-optimized implementation of the [Broyden--Fletcher--Goldfarb--Shanno optimization algorithm](https://en.wikipedia.org/wiki/Broyden%E2%80%93Fletcher%E2%80%93Goldfarb%E2%80%93Shanno_algorithm) (BFGS).
+
+BFGS is perhaps the most widely used among the so-called Quasi-Newton, second-order optimization algorithms. As opposed to the family of first-order algorithms that, in deciding on a descent direction, make use of gradient information only, second-order algorithms additionally take curvature information into account. To that end, exact Newton methods actually compute the Hessian (a costly operation), while Quasi-Newton methods avoid that cost and, instead, resort to iterative approximation.
+
+Looking at the contours of the Rosenbrock function, with its prolonged, narrow valley, it is not difficult to imagine that curvature information might make a difference. And, as you'll see in a second, it really does. Before though, one note on the code. When using L-BFGS, it is necessary to wrap function call and gradient evaluation in a closure (`calc_loss()`, in the below snippet), for them to be callable several times per iteration. You can convince yourself that the closure is, in fact, entered repeatedly, by inspecting this code snippet's chatty output:
```

Comment:
> it is necessary to wrap function call and gradient evaluation in a closure

Either give this an article (probably the), or pluralize it (function _calls_ and gradient evaluation). "function call" can't really stand alone.

Reply:
> You can convince yourself
Convince yourself sort of implies that it's not true (we convince ourselves of many false things!). Maybe try something like _You can see for yourself that the closure is…_?

Reply:
> Either give this an article (probably the), or pluralize it (function calls and gradient evaluation).

I always dislike the many "the"s I have in my English writing (they feel unidiomatic to me)... but I just had an idea: "both" should do it...

Reply:
> Convince yourself sort of implies that it's not true (we convince ourselves of many false things!). 

Honestly, I wouldn't go that far... I mean, you're right, but to me it's pretty synonymous to your alternative ...